drivers/pca9685/pca9685.c

```diff
@@ -0,0 +1,382 @@
+
+static void _set_reg_bit(uint8_t *byte, uint8_t mask, uint8_t bit);
+
+static int _read(const pca9685_t *dev, uint8_t reg, uint8_t *data, uint32_t len);
+static int _write(const pca9685_t *dev, uint8_t reg, const uint8_t *data, uint32_t len);
+static int _update(const pca9685_t *dev, uint8_t reg, uint8_t mask, uint8_t data);
+
+inline static int _write_word(const pca9685_t *dev, uint8_t reg, uint16_t word);
+
+int pca9685_init(pca9685_t *dev, const pca9685_params_t *params)
+{
+    /* some parameter sanity checks */
+    ASSERT_PARAM(dev != NULL);
+    ASSERT_PARAM(params != NULL);
+    ASSERT_PARAM(params->ext_freq <= 50000000);
+
+    dev->powered_on = false;
+    dev->params = *params;
+
+    DEBUG_DEV("params=%p", dev, params);
+
+    if (dev->params.oe_pin != GPIO_UNDEF) {
+        /* init the pin an disable outputs first */
+        gpio_init(dev->params.oe_pin, GPIO_OUT);
+        gpio_set(dev->params.oe_pin);
+    }
+
+    /* test whether PWM device is available */
+    EXEC_RET(_is_available(dev));
+
+    /* init the PWM device */
+    EXEC_RET(_init(dev));
+
+    return PCA9685_OK;
+}
+
+uint32_t pca9685_pwm_init(pca9685_t *dev, pwm_mode_t mode, uint32_t freq,
+                                                           uint16_t res)
+{
+    /* some parameter sanity checks */
+    ASSERT_PARAM(dev != NULL);
+    ASSERT_PARAM(freq >= 24 && freq <= 1526);
+    ASSERT_PARAM(res >= 2 && res <= 4096);
+    ASSERT_PARAM(mode == PWM_LEFT || mode == PWM_CENTER || mode == PWM_RIGHT);
+
+    DEBUG_DEV("mode=%u freq=%"PRIu32" res=%u", dev, mode, freq, res);
+
+    dev->params.mode = mode;
+    dev->params.freq = freq;
+    dev->params.res = res;
+
+    /* prescale can only be set while in sleep mode (powered off) */
+    if (dev->powered_on) {
+        pca9685_pwm_poweroff(dev);
+    }
+
+    /* prescale = round(clk / (PCA9685_RESOLUTION * freq)) - 1; */
+    float clk = (float)(dev->params.ext_freq ? dev->params.ext_freq
+                                             : PCA9685_OSC_FREQ);
+    float prescale = clk / PCA9685_RESOLUTION / freq;
+    uint8_t byte = round(prescale) - 1;
+    EXEC_RET(_write(dev, PCA9685_REG_PRE_SCALE, &byte, 1));
+
+    if (!dev->powered_on) {
+        pca9685_pwm_poweron(dev);
+    }
+
+    return freq;
+}
+
+void pca9685_pwm_set(pca9685_t *dev, uint8_t chn, uint16_t val)
+{
+    ASSERT_PARAM(dev != NULL);
+    ASSERT_PARAM(chn <= PCA9685_CHANNEL_NUM);
+
+    DEBUG_DEV("chn=%u val=%u", dev, chn, val);
+
+    /* limit val to resolution */
+    val = (val >= dev->params.res) ?  dev->params.res : val;
+
+    uint16_t on;
+    uint16_t off;
+
+    if (val == 0) {
+        /* full off */
+        on = 0;
+        off = PCA9685_LED_OFF;
+    }
+    else if (val == dev->params.res) {
+        /* full on */
+        on = PCA9685_LED_ON;
+        off = 0;
+    }
+    else {
+        /* duty = scale * val, with scale = 4096 / resolution; */
+        uint16_t duty = round((float)PCA9685_RESOLUTION / dev->params.res * val);
```

Comment:
```suggestion
        uint32_t div = dev->params.res * val;
        uint16_t duty = (PCA9685_RESOLUTION + div/2) / div;
```

Reply:
Ok.

Reply:
OK

Reply:
Hm, your suggestion is not correct. `duty = round(PCA9685_RESOLUTION / dev->params.res * val)`, not `duty = round(PCA9685_RESOLUTION / (dev->params.res * val))`

Reply:
I guess
```
uint32_t duty = PCA9685_RESOLUTION * val / dev->params.res
```
should make the trick. Due to the `uint32_t` the multiplication can't overflow.